release notes for `2.16.1` (#12120)
PrefectHQ/prefect: RELEASE-NOTES.md

# Prefect Release Notes

## Release 2.16.1

### TODO: !Highlight!

### Enhancements
- Add filter to prevent logging the current API key — https://github.com/PrefectHQ/prefect/pull/12072
- Add artifact data to flow run graph API — https://github.com/PrefectHQ/prefect/pull/12105

> - Update `flow.serve` to support multiple schedules — https://github.com/PrefectHQ/prefect/pull/12107

Reviewer: @bunchesofdonald if you make a highlight, would be nice if you can move to highlight section with 

```
For implementation details, see the following pull request:
    - https://github.com/PrefectHQ/prefect/pull/12107
 ```

### Fixes
- Clear runs even if deployment is paused, when updating/deleting schedules — https://github.com/PrefectHQ/prefect/pull/12089
- Surface work pool not found error — https://github.com/PrefectHQ/prefect/pull/12087

> - Ignore `schedule` in `Deployment.build_from_flow` if it's out of date — https://github.com/PrefectHQ/prefect/pull/12100

Reviewer: ```suggestion
- Ignore outdated `schedule` in `Deployment.build_from_flow — https://github.com/PrefectHQ/prefect/pull/12100
```
or maybe out-of-date

- Fix schedule instructions for `prefect deploy` — https://github.com/PrefectHQ/prefect/pull/12101
- Fix reference to `prefect deployment schedule create` — https://github.com/PrefectHQ/prefect/pull/12117

> - Allow 500s on docker test `BuildError` — https://github.com/PrefectHQ/prefect/pull/12114

Reviewer: this should have received a `maintenance` label, which would have filtered it out from being added here. we don't need to include any maintenance PRs

Reviewer: (I also just retroactively added it to that PR for 🧹 )

Author: removed in [2ae8d53](https://github.com/PrefectHQ/prefect/pull/12120/commits/2ae8d531f7d7306970ac4efad8b6149f5bcefd91)

- Fix reference to `prefect deployment schedule create` — https://github.com/PrefectHQ/prefect/pull/12117
- Allow 500s on docker test `BuildError` — https://github.com/PrefectHQ/prefect/pull/12114
- Ensure only scheduled runs can be marked late — https://github.com/PrefectHQ/prefect/pull/12113

### Documentation
- Update outdated automations concepts page image — https://github.com/PrefectHQ/prefect/pull/12059
- Update automations concept page for recent triggers and actions — https://github.com/PrefectHQ/prefect/pull/12074
- Add clarifications to tutorial and getting started — https://github.com/PrefectHQ/prefect/pull/12077
- Add minimum kubernetes version to worker guide — https://github.com/PrefectHQ/prefect/pull/12095
- Add Coiled to integrations catalog docs page — https://github.com/PrefectHQ/prefect/pull/12098
- Fix formatting on webhooks page — https://github.com/PrefectHQ/prefect/pull/12088

### Experimental
- Add feature flag for flow run infra overrides — https://github.com/PrefectHQ/prefect/pull/12065

## New Contributors
* @jrbourbeau made their first contribution in https://github.com/PrefectHQ/prefect/pull/12098

> ### Contributors
> - @jrbourbeau
> 

Reviewer: ```suggestion
```

lines 32-33 should cover it

> **All changes**: https://github.com/PrefectHQ/prefect/compare/2.16.0...preview

Reviewer: ```suggestion
**All changes**: https://github.com/PrefectHQ/prefect/compare/2.16.0...2.16.1
```

We've also added support for multiple schedules to `flow.serve`, `flow.deploy` and `prefect.runner.serve`. You can provide multiple schedules by passing a list to the `cron`, `interval`, or `rrule` arguments:

```python
import datetime
import random

from prefect import flow

@flow
def trees():
    tree = random.choice(["🌳", "🌴", "🌲", "🌵"])
    print(f"Here's a happy little tree: {tree}")

if __name__ == "__main__":
    trees.serve(
        name="trees",
        interval=[3600, 7200, 14400],
    )
```

This will create a deployment with three schedules, one that runs every hour, one that runs every two hours, and one that runs every four hours. For more advanced cases, use the `schedules` argument. 

```python
    trees.serve(
        name="trees",
        schedules=[
            IntervalSchedule(interval=datetime.timedelta(minutes=30)),
            {"schedule": RRuleSchedule(rrule="FREQ=YEARLY"), "active": True},
            MinimalDeploymentSchedule(schedule=CronSchedule(cron="0 0 * * *"), active=False),
        ]
    )
```

Dive into these new scheduling capabilities today and streamline your workflows like never before. 

> 

Author: ```suggestion
For implementation details, see the following pull request: - https://github.com/PrefectHQ/prefect/pull/12107

```

We've also added support for multiple schedules to `flow.serve`, `flow.deploy` and `prefect.runner.serve`. You can provide multiple schedules by passing a list to the `cron`, `interval`, or `rrule` arguments:

```python
import datetime
import random

from prefect import flow

@flow
def trees():
    tree = random.choice(["🌳", "🌴", "🌲", "🌵"])
    print(f"Here's a happy little tree: {tree}")

if __name__ == "__main__":
    trees.serve(
        name="trees",
        interval=[3600, 7200, 14400],
    )
```

This will create a deployment with three schedules, one that runs every hour, one that runs every two hours, and one that runs every four hours. For more advanced cases, use the `schedules` argument. 

```python
    trees.serve(
        name="trees",
        schedules=[
            IntervalSchedule(interval=datetime.timedelta(minutes=30)),
            {"schedule": RRuleSchedule(rrule="FREQ=YEARLY"), "active": True},
            MinimalDeploymentSchedule(schedule=CronSchedule(cron="0 0 * * *"), active=False),
        ]
    )
```

Dive into these new scheduling capabilities today and streamline your workflows like never before. 

> For implementation details, see the following pull request: - https://github.com/PrefectHQ/prefect/pull/12107

Reviewer 2: ```suggestion
For implementation details, see the following pull request: https://github.com/PrefectHQ/prefect/pull/12107
```

Reviewer 2: I think we can drop the dash OR use a single bullet list for the highlight. We've gone back and forth on convention in release notes.

Author: thanks! updated in [f8ea26e](https://github.com/PrefectHQ/prefect/pull/12120/commits/f8ea26efe735c64d91b2d469a2f836a308ade136)

### Enhancements

> - Add filter to prevent logging the current API key — https://github.com/PrefectHQ/prefect/pull/12072

Reviewer 2: ```suggestion
- Add a logging filter to prevent logging the current API key — https://github.com/PrefectHQ/prefect/pull/12072
```

- Update `flow.serve` to support multiple schedules — https://github.com/PrefectHQ/prefect/pull/12107
- Update `prefect deploy` to support multiple schedules — https://github.com/PrefectHQ/prefect/pull/12121

### Fixes

> - Clear runs even if deployment is paused, when updating/deleting schedules — https://github.com/PrefectHQ/prefect/pull/12089

Reviewer 2: ```suggestion
- Clear runs when updating or deleting schedules, even if the deployment is paused — https://github.com/PrefectHQ/prefect/pull/12089
```

> - Surface work pool not found error — https://github.com/PrefectHQ/prefect/pull/12087

Reviewer 2: ```suggestion
- Surface missing work pool errors when working with pools in the CLI — https://github.com/PrefectHQ/prefect/pull/12087
```

# Prefect Release Notes

## Release 2.16.1

### Enhanced multiple schedule support  

`prefect.yaml` now supports specifying multiple schedules via the `schedules` key. This allows you to define multiple schedules for a single deployment, and each schedule can have its own `cron`, `interval`, or `rrule` configuration:

```yaml
 ...
 schedules:
    - cron: "0 0 * * *"
      active: false
    - interval: 3600
      active: true
    - rrule: "FREQ=YEARLY"
      active: true
```

> In addition you can specify multiple schedules via arguments to `prefect deploy`:

Reviewer: ```suggestion
In addition, you can now specify multiple schedules via arguments to `prefect deploy`:
```

```python
    trees.serve(
        name="trees",
        schedules=[
            IntervalSchedule(interval=datetime.timedelta(minutes=30)),
            {"schedule": RRuleSchedule(rrule="FREQ=YEARLY"), "active": True},
            MinimalDeploymentSchedule(schedule=CronSchedule(cron="0 0 * * *"), active=False),
        ]
    )
```

Dive into these new scheduling capabilities today and streamline your workflows like never before. 

For implementation details, see the following pull request:
    - https://github.com/PrefectHQ/prefect/pull/12107

### Enhancements
- Add a logging filter to prevent logging the current API key — https://github.com/PrefectHQ/prefect/pull/12072
- Update `flow.serve` to support multiple schedules — https://github.com/PrefectHQ/prefect/pull/12107
- Update `prefect deploy` to support multiple schedules — https://github.com/PrefectHQ/prefect/pull/12121

### Fixes
- Clear runs when updating or deleting schedules, even if the deployment is paused — https://github.com/PrefectHQ/prefect/pull/12089

> - Surface missing work pool errors when working with pools in the CLI — https://github.com/PrefectHQ/prefect/pull/12087

Reviewer: ```suggestion
- Surface missing work pool error in CLI — https://github.com/PrefectHQ/prefect/pull/12087
```

> - Ignore outdated `schedule` in `Deployment.build_from_flow — https://github.com/PrefectHQ/prefect/pull/12100

Reviewer: ```suggestion
- Ignore outdated `schedule` in `Deployment.build_from_flow` — https://github.com/PrefectHQ/prefect/pull/12100
```

- Fix schedule instructions for `prefect deploy` — https://github.com/PrefectHQ/prefect/pull/12101
- Fix reference to `prefect deployment schedule create` — https://github.com/PrefectHQ/prefect/pull/12117
- Ensure only scheduled runs can be marked late — https://github.com/PrefectHQ/prefect/pull/12113

### Documentation
- Update outdated automations concepts page image — https://github.com/PrefectHQ/prefect/pull/12059
- Update automations concept page for recent triggers and actions — https://github.com/PrefectHQ/prefect/pull/12074
- Add clarifications to tutorial and getting started pages — https://github.com/PrefectHQ/prefect/pull/12077

> - Add minimum kubernetes version to worker guide — https://github.com/PrefectHQ/prefect/pull/12095

Reviewer: ```suggestion
- Add minimum Kubernetes version to worker guide — https://github.com/PrefectHQ/prefect/pull/12095
```
